pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -0,0 +1,394 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	adsDebug = os.Getenv("PILOT_DEBUG_ADS") != "0"
+
+	// adsClients reflect active gRPC channels, for both ADS and EDS.
+	adsClients      = map[string]*XdsConnection{}
```

Comment:
Please encapsulate this in a struct and use that struct on the methods that need to access this data (i.e. put this into the DiscoveryServer). As-is it's impossible to mock these connections, and the global variable is brittle. Finally, we should just use a sync.Map here and not a map+mutex, as I said the first go-round in the EDS implementation.

Reply:
I don't think either is critical for this PR - we're not trying to mock the connections and performance-wise there is not a lot of concurrent access, and I think simpler is better in this case. 

We can refine it in future if needed.